README.md
# AutoMongoBackup

This is a very barebones port of the [AutoMySQLBackup](http://sourceforge.net/projects/automysqlbackup) project. This bash script will allow you to do daily backups of a Mongo database. It includes periodic rotation so that you can keep historical weekly and monthly backups, as well as daily.

## One db, all collections

    DBNAME="mydb"

## One db, one collection

    DBNAME="mydb"
    COLNAME= "mycollection"

## License

see [CHANGES.md](CHANGES.md) for authors

This program is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public License
as published by the Free Software Foundation; either version 2
of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program; if not, write to the Free Software
Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.


CHANGELOG.md
* VER 0.9.3 - (2014-12-24) (author: Omer Akhter)
  * Added xz compression option
  * Added option to log to single file
  * Added option to keep logs
* VER 0.9.2 - (2014-xx-xx) (author: Alex Bevilacqua)
  * Added 7zip compression option
* VER 0.9.1 - (2014-xx-xx) (author: Xisco Guaita)
  * Added individual database backup
  * Added individual collection backup
* VER 0.9 - (2011-10-28) (author: Joshua Keroes)
  * Fixed bugs and improved logic in select_secondary_member()
  * Fixed minor grammar issues and formatting in docs
* VER 0.8 - (2011-10-02) (author: Krzysztof Wilczynski)
  * Added better support for selecting Secondary member in the Replica Sets that can be used to take backups without bothering busy Primary member too much.
* VER 0.7 - (2011-09-23) (author: Krzysztof Wilczynski)
  * Added support for --journal dring taking backup to enable journaling.
* VER 0.6 - (2011-09-15) (author: Krzysztof Wilczynski)
  * Added support for --oplog during taking backup for point-in-time snapshotting.
  * Added filter for "mongodump" writing "connected to:" on the standard error, which is not desirable.
* VER 0.5 - (2011-02-04) (author: Jan Doberstein)
  * Added replicaset support (don't Backup on Master)
  * Added Hard Support for 'latest' Copy
* VER 0.4 - (2010-10-26)
      * Cleaned up warning message to make it clear that it can usually be safely ignored
* VER 0.3 - (2010-06-11)
  * Added the DBPORT parameter
  * Changed USERNAME and PASSWORD to DBUSERNAME and DBPASSWORD
  * Fixed some bugs with compression
* VER 0.2 - (2010-05-27) (author: Gregory Barchard)
  * Added back the compression option for automatically creating tgz or bz2 archives
  * Added a cleanup option to optionally remove the database dump after creating the archives
  * Removed unnecessary path additions
* VER 0.1 - (2010-05-11)
  * Initial Release


src/automongobackup.sh
#!/bin/bash
#
# MongoDB Backup Script
# VER. 0.9.3
# More Info: https://github.com/omer-akhter/automongobackup
#   original source: http://github.com/dacgrouplabs/automongobackup

# Note, this is a lobotomized port of AutoMySQLBackup
# (http://sourceforge.net/projects/automysqlbackup/) for use with
# MongoDB.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
#
#=====================================================================
#=====================================================================
# Set the following variables to your system needs
# (Detailed instructions below variables)
#=====================================================================

# Username to access the mongo server e.g. dbuser
# Unnecessary if authentication is off
# DBUSERNAME=""

# Password to access the mongo server e.g. password
# Unnecessary if authentication is off
# DBPASSWORD=""

# Database for authentication to the mongo server e.g. admin
# Unnecessary if authentication is off
# DBAUTHDB=""

# Specifies a database to backup. If you do not specify a database,
# mongodump copies all databases in this instance into the dump files
# Note that this option can not be used with OPLOG="yes"
# DBNAME=

# Specifies a collection to backup. If you do not specify a collection,
# this option copies all collections in the specified database or instance
# COLNAME=

# Prefix to prepend to the backup file name
FILEPREFIX=""

# Host name (or IP address) of mongo server e.g localhost
DBHOST="127.0.0.1"

# Port that mongo is listening on
DBPORT="27017"

# Backup directory location e.g /backups
BACKUPDIR="/data/backup"

# Mail setup
# What would you like to be mailed to you?
# - log   : send only log file
# - files : send log file and sql files as attachments (see docs)
# - stdout : will simply output the log to the screen if run manually.
# - quiet : Only send logs if an error occurs to the MAILADDR.
MAILCONTENT="stdout"

# Set the maximum allowed email size in k. (4000 = approx 5MB email [see docs])
MAXATTSIZE="4000"

# Email Address to send mail to? (user@domain.com)
# MAILADDR=""

# ============================================================
# === ADVANCED OPTIONS ( Read the doc's below for details )===
#=============================================================

# Which day do you want weekly backups? (1 to 7 where 1 is Monday)
DOWEEKLY=6

# Choose Compression type. (7z, gzip, bzip2 or xz)
COMP="7z"

# Choose if the uncompressed folder should be deleted after compression has completed
CLEANUP="yes"

# Additionally keep a copy of the most recent backup in a seperate directory.
LATEST="yes"

# Make Hardlink not a copy
LATESTLINK="yes"

# Use oplog for point-in-time snapshotting.
OPLOG="no"

# Choose other Server if is Replica-Set Master
REPLICAONSLAVE="yes"

# Command to run before backups (uncomment to use)
# PREBACKUP=""

# Command run after backups (uncomment to use)
# POSTBACKUP=""

# Should the log files be deleted
#CLEANUP_LOG="yes"

# Should the error log files be deleted
#CLEANUP_ERR="yes"

# Log to single file
SINGLE_LOG="yes"

#=====================================================================
# Options documentation
#=====================================================================
# Set DBUSERNAME and DBPASSWORD of a user that has at least SELECT permission
# to ALL databases. Alternatively set DBNAME and provide credentials to that DB
#
# Set the DBHOST option to the server you wish to backup, leave the
# default to backup "this server".(to backup multiple servers make
# copies of this file and set the options for that server)
#
# You can change the backup storage location from /backups to anything
# you like by using the BACKUPDIR setting..
#
# The MAILCONTENT and MAILADDR options and pretty self explanatory, use
# these to have the backup log mailed to you at any email address or multiple
# email addresses in a space seperated list.
#
# (If you set mail content to "log" you will require access to the "mail" program
# on your server. If you set this to "files" you will have to have mutt installed
# on your server. If you set it to "stdout" it will log to the screen if run from
# the console or to the cron job owner if run through cron. If you set it to "quiet"
# logs will only be mailed if there are errors reported. )
#
#
# Finally copy automongobackup.sh to anywhere on your server and make sure
# to set executable permission. You can also copy the script to
# /etc/cron.daily to have it execute automatically every night or simply
# place a symlink in /etc/cron.daily to the file if you wish to keep it
# somwhere else.
#
# NOTE: On Debian copy the file with no extention for it to be run
# by cron e.g just name the file "automongobackup"
#
# Thats it..
#
#
# === Advanced options ===
#
# To set the day of the week that you would like the weekly backup to happen
# set the DOWEEKLY setting, this can be a value from 1 to 7 where 1 is Monday,
# The default is 6 which means that weekly backups are done on a Saturday.
#
# Use PREBACKUP and POSTBACKUP to specify Pre and Post backup commands
# or scripts to perform tasks either before or after the backup process.
#
#
#=====================================================================
# Backup Rotation..
#=====================================================================
#
# Daily Backups are rotated weekly.
#
# Weekly Backups are run by default on Saturday Morning when
# cron.daily scripts are run. This can be changed with DOWEEKLY setting.
#
# Weekly Backups are rotated on a 5 week cycle.
# Monthly Backups are run on the 1st of the month.
# Monthly Backups are NOT rotated automatically.
#
# It may be a good idea to copy Monthly backups offline or to another
# server.
#
#=====================================================================
# Please Note!!
#=====================================================================
#
# I take no resposibility for any data loss or corruption when using
# this script.
#
# This script will not help in the event of a hard drive crash. You
# should copy your backups offline or to another PC for best protection.
#
# Happy backing up!
#
#=====================================================================
# Restoring
#=====================================================================
# ???
#
#=====================================================================
# Change Log
#=====================================================================
# VER 0.9.3 - (2014-12-24) (author: Omer Akhter)
#   - Added xz compression option
#   - Added option to log to single file
#   - Added option to keep logs
# VER 0.9.2 - (2014-xx-xx) (author: Alex Bevilacqua)
#	- Added 7zip compression option
# VER 0.9.1 - (2014-xx-xx) (author: Xisco Guaita)
#       - Added individual database backup
#       - Added individual collection backup
# VER 0.9 - (2011-10-28) (author: Joshua Keroes)
#       - Fixed bugs and improved logic in select_secondary_member()
#       - Fixed minor grammar issues and formatting in docs
#
# VER 0.8 - (2011-10-02) (author: Krzysztof Wilczynski)
#       - Added better support for selecting Secondary member in the
#         Replica Sets that can be used to take backups without bothering
#         busy Primary member too much.
#
# VER 0.7 - (2011-09-23) (author: Krzysztof Wilczynski)
#       - Added support for --journal dring taking backup
#         to enable journaling.
#
# VER 0.6 - (2011-09-15) (author: Krzysztof Wilczynski)
#       - Added support for --oplog during taking backup for
#         point-in-time snapshotting.
#       - Added filter for "mongodump" writing "connected to:"
#         on the standard error, which is not desirable.
#
# VER 0.5 - (2011-02-04) (author: Jan Doberstein)
#       - Added replicaset support (don't Backup on Master)
#       - Added Hard Support for 'latest' Copy
#
# VER 0.4 - (2010-10-26)
#       - Cleaned up warning message to make it clear that it can
#         usually be safely ignored
#
# VER 0.3 - (2010-06-11)
#       - Added the DBPORT parameter
#       - Changed USERNAME and PASSWORD to DBUSERNAME and DBPASSWORD
#       - Fixed some bugs with compression
#
# VER 0.2 - (2010-05-27) (author: Gregory Barchard)
#       - Added back the compression option for automatically creating
#         tgz or bz2 archives
#       - Added a cleanup option to optionally remove the database dump
#         after creating the archives
#       - Removed unnecessary path additions
#
# VER 0.1 - (2010-05-11)
#       - Initial Release
#
#=====================================================================
#=====================================================================
#=====================================================================
#
# Should not need to be modified from here down!!
#
#=====================================================================
#=====================================================================
#=====================================================================

shellout () {
    if [ -n "$1" ]; then
        echo $1
        exit 1
    fi
    exit 0
}

# External config - override default values set above
for x in default sysconfig; do
  if [ -f "/etc/$x/automongobackup" ]; then
      source /etc/$x/automongobackup
  fi
done

# Include extra config file if specified on commandline, e.g. for backuping several remote dbs from central server
[ ! -z "$1" ] && [ -f "$1" ] && source ${1}

#=====================================================================

PATH=/usr/local/bin:/usr/bin:/bin
DATE=`date +%Y-%m-%d_%Hh%Mm`                      # Datestamp e.g 2002-09-21
DOW=`date +%A`                                    # Day of the week e.g. Monday
DNOW=`date +%u`                                   # Day number of the week 1 to 7 where 1 represents Monday
DOM=`date +%d`                                    # Date of the Month e.g. 27
M=`date +%B`                                      # Month e.g January
W=`date +%V`                                      # Week Number e.g 37
VER=0.9.3                                         # Version Number

LOGFILE_PREFIX=$BACKUPDIR/$DBHOST.log
LOGERR_PREFIX=$BACKUPDIR/ERRORS_$DBHOST.log
LOGFILE=$LOGFILE_PREFIX-`date +%H%M`.log          # Logfile Name
LOGERR=$LOGERR_PREFIX-`date +%H%M`.log            # Errologfile Name

BACKUPFILES=""
OPT=""                                            # OPT string for use with mongodump

# Do we need to use a username/password?
if [ "$DBUSERNAME" ]; then
    OPT="$OPT --username $DBUSERNAME --password $DBPASSWORD --authenticationDatabase $DBAUTHDB"
fi

# Do we want one or all databases?
if [ "$DBNAME" ]; then
  OPT="$OPT --db $DBNAME"
  FILEPREFIX="$FILEPREFIX$DBNAME-"
  # Do we want one or all collections?
  if [ "$COLNAME" ]; then
    OPT="$OPT --collection  $COLNAME"
    FILEPREFIX="$FILEPREFIX$COLNAME-"
  fi
else
  # Do we use oplog for point-in-time snapshotting of all databases?
  # Can not be used for single-database dumps
  if [ "$OPLOG" = "yes"  ]; then
    OPT="$OPT --oplog"
  fi
fi

# Create required directories
mkdir -p $BACKUPDIR/{daily,weekly,monthly} || shellout 'failed to create directories'

if [ "$LATEST" = "yes" ]; then
    #rm -rf "$BACKUPDIR/latest"
    mkdir -p "$BACKUPDIR/latest" || shellout 'failed to create directory'
fi

# Check for correct sed usage
if [ $(uname -s) = 'Darwin' -o $(uname -s) = 'FreeBSD' ]; then
    SED="sed -i ''"
else
    SED="sed -i"
fi

# IO redirection for logging.
touch $LOGFILE
exec 6>&1           # Link file descriptor #6 with stdout.
                    # Saves stdout.
exec > $LOGFILE     # stdout replaced with file $LOGFILE.

touch $LOGERR
exec 7>&2           # Link file descriptor #7 with stderr.
                    # Saves stderr.
exec 2> $LOGERR     # stderr replaced with file $LOGERR.

# When a desire is to receive log via e-mail then we close stdout and stderr.
[ "x$MAILCONTENT" == "xlog" ] && exec 6>&- 7>&-

# Functions

# Database dump function
dbdump () {
    #echo mongodump --host $DBHOST:$DBPORT --out $1 $OPT
    mongodump --host $DBHOST:$DBPORT --out $1 $OPT
    [ -e "$1" ] && return 0
    echo "ERROR: mongodump failed to create dumpfile: $1" >&2
    return 1
}

#
# Select first available Secondary member in the Replica Sets and show its
# host name and port.
#
function select_secondary_member {
    # We will use indirect-reference hack to return variable from this function.
    local __return=$1

    # Return list of with all replica set members
    members=( $(mongo --quiet --host $DBHOST:$DBPORT --eval 'rs.conf().members.forEach(function(x){ print(x.host) })') )

    # Check each replset member to see if it's a secondary and return it.
    if [ ${#members[@]} -gt 1 ]; then
        for member in "${members[@]}"; do

            is_secondary=$(mongo --quiet --host $member --eval 'rs.isMaster().secondary')
            case "$is_secondary" in
                'true')     # First secondary wins ...
                    secondary=$member
                    break
                ;;
                'false')    # Skip particular member if it is a Primary.
                    continue
                ;;
                *)          # Skip irrelevant entries.  Should not be any anyway ...
                    continue
                ;;
            esac
        done
    fi

    if [ -n "$secondary" ]; then
        # Ugly hack to return value from a Bash function ...
        eval $__return="'$secondary'"
    fi
}

# Compression function plus latest copy
compression () {
    SUFFIX=""
    dir=$(dirname $1)
    file=$(basename $1)
    if [ -n "$COMP" ]; then
        if [ "$COMP" = "7z" ]; then
            # TODO check that 7zip is available
            #   ex: command -v 7z > /dev/null 2>&1 || { echo >&2 "7-Zip missing (SKIPPING_COMPRESSION)";  }
            echo 7-zipping to "$file.7z"
            cd "$dir" && 7z a "$file.7z" *
        else
            [ "$COMP" = "gzip" ] && SUFFIX=".tgz"
            [ "$COMP" = "bzip2" ] && SUFFIX=".tar.bz2"
            [ "$COMP" = "xz" ] && SUFFIX=".tar.xz"
            echo Tar and $COMP to "$file$SUFFIX"
            cd "$dir" && tar -cf - "$file" | $COMP -c > "$file$SUFFIX"
        fi
        cd - >/dev/null || return 1
    else
        echo "No compression option set, check advanced settings"
    fi

    if [ "$LATEST" = "yes" ]; then
    # only remove this prefix
    rm -f $BACKUPDIR/latest/${FILEPREFIX}*
        if [ "$LATESTLINK" = "yes" ];then
            COPY="ln"
        else
            COPY="cp"
        fi
        $COPY "$1$SUFFIX" "$BACKUPDIR/latest/"
    fi

    if [ "$CLEANUP" = "yes" ]; then
        echo Cleaning up folder at "$1"
        rm -rf "$1"
    fi

    return 0
}

# Run command before we begin
if [ "$PREBACKUP" ]; then
    echo ======================================================================
    echo "Prebackup command output."
    echo
    eval $PREBACKUP
    echo
    echo ======================================================================
    echo
fi

# Hostname for LOG information
if [ "$DBHOST" = "localhost" -o "$DBHOST" = "127.0.0.1" ]; then
    HOST=`hostname`
    if [ "$SOCKET" ]; then
        OPT="$OPT --socket=$SOCKET"
    fi
else
    HOST=$DBHOST
fi

# Try to select an available secondary for the backup or fallback to DBHOST.
if [ "x${REPLICAONSLAVE}" == "xyes" ]; then
    # Return value via indirect-reference hack ...
    select_secondary_member secondary

    if [ -n "$secondary" ]; then
        DBHOST=${secondary%%:*}
        DBPORT=${secondary##*:}
    else
        SECONDARY_WARNING="WARNING: No suitable Secondary found in the Replica Sets.  Falling back to ${DBHOST}."
    fi
fi

echo ======================================================================
echo AutoMongoBackup VER $VER

if [ ! -z "$SECONDARY_WARNING" ]; then
    echo
    echo "$SECONDARY_WARNING"
fi

echo
echo Backup of Database Server - $HOST on $DBHOST
echo ======================================================================

echo Backup Start `date`
echo ======================================================================
# Monthly Full Backup of all Databases
if [ $DOM = "01" ]; then
    echo Monthly Full Backup
    FILE="$BACKUPDIR/monthly/${FILEPREFIX}$M.$DATE"

# Weekly Backup
elif [ $DNOW = $DOWEEKLY ]; then
    echo Weekly Backup
    echo
    echo Rotating 5 weeks Backups...
    if [ "$W" -le 05 ]; then
        REMW=`expr 48 + $W`
    elif [ "$W" -lt 15 ]; then
        REMW=0`expr $W - 5`
    else
        REMW=`expr $W - 5`
    fi
    rm -f $BACKUPDIR/weekly/${FILEPREFIX}$REMW.*
    echo
    FILE="$BACKUPDIR/weekly/${FILEPREFIX}$W.$DATE"

# Daily Backup
else
    echo Daily Backup of Databases
    echo Rotating last weeks Backup...
    echo
    rm -f $BACKUPDIR/daily/${FILEPREFIX}$DOW.*
    echo
    FILE="$BACKUPDIR/daily/${FILEPREFIX}$DOW.$DATE"
fi
dbdump $FILE && compression $FILE
echo ----------------------------------------------------------------------
echo Backup End Time `date`
echo ======================================================================

echo Total disk space used for backup storage..
echo Size - Location
echo `du -hs "$BACKUPDIR"`
echo
echo ======================================================================

# Run command when we're done
if [ "$POSTBACKUP" ]; then
    echo ======================================================================
    echo "Postbackup command output."
    echo
    eval $POSTBACKUP
    echo
    echo ======================================================================
fi

# Clean up IO redirection if we plan not to deliver log via e-mail.
[ ! "x$MAILCONTENT" == "xlog" ] && exec 1>&6 2>&7 6>&- 7>&-

if [ -s "$LOGERR" ]; then
    eval $SED "/^connected/d" "$LOGERR"
fi

if [ "$MAILCONTENT" = "log" ]; then
    cat "$LOGFILE" | mail -s "Mongo Backup Log for $HOST - $DATE" $MAILADDR

    if [ -s "$LOGERR" ]; then
        cat "$LOGERR"
        cat "$LOGERR" | mail -s "ERRORS REPORTED: Mongo Backup error Log for $HOST - $DATE" $MAILADDR
    fi
else
    if [ -s "$LOGERR" ]; then
        cat "$LOGFILE"
        echo
        echo "###### WARNING ######"
        echo "STDERR written to during mongodump execution."
        echo "The backup probably succeeded, as mongodump sometimes writes to STDERR, but you may wish to scan the error log below:"
        cat "$LOGERR"
    else
        cat "$LOGFILE"
    fi
fi

# TODO: Would be nice to know if there were any *actual* errors in the $LOGERR
STATUS=0
if [ -s "$LOGERR" ]; then
    STATUS=1
fi

# Clean up Logfile
if [ "x${SINGLE_LOG}" == "xyes"]; then
	if [ "x${CLEANUP_LOG}" == "xyes" ]; then
		[ -f "$LOGFILE_PREFIX" ] && rm -fv "$LOGFILE_PREFIX"
	else
		echo >> "$LOGFILE_PREFIX"
		echo "#####################" >> "$LOGFILE_PREFIX"
		echo "# $DATE #" >> "$LOGFILE_PREFIX"
		echo "#####################" >> "$LOGFILE_PREFIX"
		cat "$LOGFILE" >> "$LOGFILE_PREFIX"
	fi
	rm -fv "$LOGFILE"

	if [ "x${CLEANUP_ERR}" == "xyes" ]; then
		[ -f "$LOGERR_PREFIX" ] && rm -fv "$LOGERR_PREFIX"
	else
		echo >> "$LOGERR_PREFIX"
		echo "#####################" >> "$LOGERR_PREFIX"
		echo "# $DATE #" >> "$LOGERR_PREFIX"
		echo "#####################" >> "$LOGERR_PREFIX"
		cat "$LOGERR" >> "$LOGERR_PREFIX"
	fi
	rm -fv "$LOGERR"
else
	[ "x${CLEANUP_LOG}" == "xyes" ] && [ -f "$LOGFILE" ] && rm -fv "$LOGFILE"
	[ "x${CLEANUP_ERR}" == "xyes" ] && [ -f "$LOGERR" ] && rm -fv "$LOGERR"
fi

exit $STATUS
